### PR TITLE
1615 jaxrs part 2

### PR DESCRIPTION
**Pull request by xsalefter**
Some notes:

- [JaxRsResourceBase.deleteCustomFields()](https://github.com/killbill/killbill/blob/cef3351c4ec44347d478f61e420ae8d443fc62f8/jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/JaxRsResourceBase.java#L300), add small logic `customFieldList == null` to match specs in javadoc, " .... or null if they should all be removed"

- Move `TestJaxRsResourceBase.testGetHighestRecordDate()` to `TestUsageResource.testGetHighestRecordDate()`

- extract `BundleResource.getBundles()` and `BundleResource.searchBundles()` methods to `buildBundleStreamPaginationResponse()` because they are contains (almost) the same method body.

### CODE REVIEW

**Comment by xsalefter** [comment]
@sbrossie @pierre failed because problem described in image below. Should we ignore this? Or perhaps change regex in `DefaultTagUserApi.createTagDefinition()` to something like this:

```java
    public static boolean containsUpperCase(final String s) {
        if (isNullOrEmpty(s)) {
            return false; // same as regex ".*[A-Z].*"
        }
        return !s.equals(s.toLowerCase());
    }
```


Screenshot from 2022-06-08 07-53-51

**Comment by sbrossie** [comment]
> @sbrossie @pierre failed because problem described in image below. Should we ignore this? Or perhaps change regex in `DefaultTagUserApi.createTagDefinition()` to something like this:
> 
> ```java
>     public static boolean containsUpperCase(final String s) {
>         if (isNullOrEmpty(s)) {
>             return false; // same as regex ".*[A-Z].*"
>         }
>         return !s.equals(s.toLowerCase());
>     }
> ```
> 
> Screenshot from 2022-06-08 07-53-51

@xsalefter If the concern is that the input string may be too long and run into `n^k`, we could first check that the string is not too long - it should not be longer than what our [schema](https://github.com/killbill/killbill/blob/work-for-release-0.23.x/util/src/main/resources/org/killbill/billing/util/ddl.sql#L53) allows, and then keep the regex as-is, as there would be no more concern.

File: jaxrs/src/test/java/org/killbill/billing/jaxrs/resources/TestJaxRsResourceBase.java [hunk 1]
```diff
@@ -17,31 +17,58 @@
...
 import org.testng.annotations.Test;
 
-import com.google.common.collect.ImmutableList;
-
 public class TestJaxRsResourceBase extends JaxrsTestSuiteNoDB {
 
-    private final JaxRsResourceBaseTest base = new JaxRsResourceBaseTest();
+    private CustomFieldUserApi customFieldUserApi;
+
+    @BeforeMethod(groups = "fast", alwaysRun = true)
```
Comment (pierre) [inline jaxrs/src/test/java/org/killbill/billing/jaxrs/resources/TestJaxRsResourceBase.java:42]: No need for `alwaysRun = true`.

File: jaxrs/src/test/java/org/killbill/billing/jaxrs/resources/TestJaxRsResourceBase.java [hunk 2]
```diff
@@ -17,31 +17,58 @@
...
+    public void beforeMethod() {
+        customFieldUserApi = Mockito.mock(CustomFieldUserApi.class);
```
Comment (pierre) [inline jaxrs/src/test/java/org/killbill/billing/jaxrs/resources/TestJaxRsResourceBase.java:44]: Is there a `super` method to call?

Regardless, I think you also need:
```
        if (hasFailed()) {
            return;
        }
```
to work around issues with TestNG where it runs these methods even if the suite is failing.

**Review by pierre — APPROVED** [review]
Just a couple of comments, but otherwise :shipit:

**Comment by pierre** [comment]
> @sbrossie @pierre failed because problem described in image below. Should we ignore this? Or perhaps change regex in `DefaultTagUserApi.createTagDefinition()` to something like this:
> 
> ```java
>     public static boolean containsUpperCase(final String s) {
>         if (isNullOrEmpty(s)) {
>             return false; // same as regex ".*[A-Z].*"
>         }
>         return !s.equals(s.toLowerCase());
>     }
> ```

Agreed to this ☝️ Let's just change the implementation.

File: jaxrs/src/test/java/org/killbill/billing/jaxrs/resources/TestJaxRsResourceBase.java (hunk 2 above)
Comment (xsalefter) [reply jaxrs/src/test/java/org/killbill/billing/jaxrs/resources/TestJaxRsResourceBase.java:47]: add `if (hasFailed()) { return; }` works.

**Comment by xsalefter** [comment]
@sbrossie I took @pierre 's [approach](https://github.com/killbill/killbill/pull/1712#issuecomment-1149506140) instead of yours :sweat_smile: .

**Comment by sbrossie** [comment]
> @sbrossie I took @pierre 's [approach](https://github.com/killbill/killbill/pull/1712#issuecomment-1149506140) instead of yours 😅 .

All good!

**Pull request merged**